libs/community/langchain_community/vectorstores/lantern.py

```diff
@@ -0,0 +1,851 @@
+        cmetadata = sqlalchemy.Column(JSON, nullable=True)
+        # custom_id : any user defined id
+        custom_id = sqlalchemy.Column(sqlalchemy.String, nullable=True)
+        embedding = sqlalchemy.Column(sqlalchemy.ARRAY(embedding_type))  # type: ignore
+
+    return EmbeddingStore
+
+
+class QueryResult:
+    """Result from a query."""
+
+    EmbeddingStore: BaseEmbeddingStore
+    distance: float
+
+
+class DistanceStrategy(str, enum.Enum):
+    """Enumerator of the Distance strategies."""
+
+    EUCLIDEAN = "l2sq"
+    COSINE = "cosine"
+    HAMMING = "hamming"
+
+
+DEFAULT_DISTANCE_STRATEGY = DistanceStrategy.COSINE
+
+
+class Lantern(VectorStore):
+    """`Postgres` with the `lantern` extension as a vector store.
+
+    lantern uses sequential scan by default. but you can create a HNSW index
+    using the create_hnsw_index method.
+    - `connection_string` is a postgres connection string.
+    - `embedding_function` any embedding function implementing
+        `langchain.embeddings.base.Embeddings` interface.
+    - `collection_name` is the name of the collection to use. (default: langchain)
+        - NOTE: This is the name of the table in which embedding data will be stored
+            The table will be created when initializing the store (if not exists)
+            So, make sure the user has the right permissions to create tables.
+    - `distance_strategy` is the distance strategy to use. (default: EUCLIDEAN)
+        - `EUCLIDEAN` is the euclidean distance.
+        - `COSINE` is the cosine distance.
+        - `HAMMING` is the hamming distance.
+    - `pre_delete_collection` if True, will delete the collection if it exists.
+        (default: False)
+        - Useful for testing.
+    """
+
+    def __init__(
+        self,
+        connection_string: str,
+        embedding_function: Embeddings,
+        distance_strategy: DistanceStrategy = DEFAULT_DISTANCE_STRATEGY,
+        collection_name: str = _LANGCHAIN_DEFAULT_COLLECTION_NAME,
+        collection_metadata: Optional[dict] = None,
+        pre_delete_collection: bool = False,
+        logger: Optional[logging.Logger] = None,
+        relevance_score_fn: Optional[Callable[[float], float]] = None,
+    ) -> None:
+        self.connection_string = connection_string
+        self.embedding_function = embedding_function
+        self.collection_name = collection_name
+        self.collection_metadata = collection_metadata
+        self._distance_strategy = distance_strategy
+        self.pre_delete_collection = pre_delete_collection
+        self.logger = logger or logging.getLogger(__name__)
+        self.override_relevance_score_fn = relevance_score_fn
+        self.EmbeddingStore = get_embedding_store(
+            self.distance_strategy, collection_name
+        )
+        self.__post_init__()
+
+    def __post_init__(
+        self,
+    ) -> None:
+        self._conn = self.connect()
+        self.create_hnsw_extension()
```

Comment:
Ideally any communication-heavy/blocking operations happen in a class method constructor rather than in the `__init__` or post init methods, but it does work so not a blocker to land this.

Reply:
Got it! `create_hnsw_extension` should not take much time here, as we don't have any heavy blocking operations in our extension creation logic.